Lib/test/test_capi/test_number.py

```diff
@@ -0,0 +1,612 @@
+
+        self.assertRaises(TypeError, multiply, 1, object())
+        self.assertRaises(TypeError, multiply, object(), 1)
+        self.assertRaises(TypeError, multiply, ndarray([1], (1,)), 2)
+        self.assertRaises(TypeError, multiply, 2, ndarray([1], (1,)))
+        self.assertRaises(TypeError, multiply, [1], 3.14)
+        self.assertRaises(OverflowError, multiply, [1], PY_SSIZE_T_MAX + 1)
+
+        # CRASHES multiply(NULL, 42)
+        # CRASHES multiply(42, NULL)
+
+    def test_matrixmultiply(self):
+        # Test PyNumber_MatrixMultiply()
+        matrixmultiply = _testcapi.number_matrixmultiply
+
+        a, b, r = WithMatrixMul([1]), 2, [1, 1]
+        self.assertEqual(matrixmultiply(a, b), r)
+        self.assertEqual(a, [1])
+        self.assertEqual(b, 2)
+
+        # CRASHES matrixmultiply(NULL, NULL)
+
+    def test_floordivide(self):
+        # Test PyNumber_FloorDivide()
+        floordivide = _testcapi.number_floordivide
+
+        self.assertEqual(floordivide(4, 3), 4 // 3)
+
+        # CRASHES floordivide(NULL, 42)
+        # CRASHES floordivide(42, NULL)
+
+    def test_truedivide(self):
+        # Test PyNumber_TrueDivide()
+        truedivide = _testcapi.number_truedivide
+
+        self.assertEqual(truedivide(3, 4), 3 / 4)
+
+        # CRASHES truedivide(NULL, 42)
+        # CRASHES truedivide(42, NULL)
+
+    def test_remainder(self):
+        # Test PyNumber_Remainder()
+        remainder = _testcapi.number_remainder
+
+        self.assertEqual(remainder(4, 3), 4 % 3)
+
+        # CRASHES remainder(NULL, 42)
+        # CRASHES remainder(42, NULL)
+
+    def test_divmod(self):
+        # Test PyNumber_divmod()
+        divmod_ = _testcapi.number_divmod
+
+        self.assertEqual(divmod_(4, 3), divmod(4, 3))
+
+        # CRASHES divmod_(NULL, 42)
+        # CRASHES divmod_(42, NULL)
+
+    def test_power(self):
+        # Test PyNumber_Power()
+        power = _testcapi.number_power
+
+        self.assertEqual(power(4, 3), pow(4, 3))
+        self.assertEqual(power(0.5, 2), pow(0.5, 2))
+        self.assertEqual(power(2, -1.0), pow(2, -1.0))
+        self.assertEqual(power(4, 11, 5), pow(4, 11, 5))
+        self.assertEqual(power(4, IntSubclass(11)), pow(4, 11))
+        self.assertEqual(power(4, IntSubclass2(11)), 11)
+        self.assertEqual(power(4, IntSubclass2(NotImplemented), NULL), 1)
+
+        self.assertRaises(TypeError, power, "spam", 42)
+        self.assertRaises(TypeError, power, object(), 42)
+        self.assertRaises(TypeError, power, 42, "spam")
+        self.assertRaises(TypeError, power, 42, object())
+        self.assertRaises(TypeError, power, 1, 2, "spam")
+        self.assertRaises(TypeError, power, 1, 2, object())
+
+        # CRASHES power(NULL, 42)
+        # CRASHES power(42, NULL)
+
+    def test_negative(self):
+        # Test PyNumber_Negative()
+        negative = _testcapi.number_negative
+
+        self.assertEqual(negative(42), -42)
```

Comment:
For all tests currently with integers only, add also tests with floats and mixed.

Reply:
Did this with unary ops, then will extend to other tests as well.

Reply:
In fact, I think the old test layout was a bad idea.  In the last commit I'm using just one method to test all four unary functions (PyNumber_Negative and friends).

This should be doable for binary and ternary functions as well, so I'll go that way, probably could do pr ready tomorrow.  Special handling will be required for PyNumber_*Add/Multiply, which also utilize tp_as_sequence.  Here almost all cases could be tested using built-in types.  But not all.  Thus, I think some C-coded test type must be used.